File: app/src/main/java/com/example/harry/customandroid/tabs/develop/customView/CanvasView.java
package com.example.harry.customandroid.tabs.develop.customView;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;

public class CanvasView extends AppCompatTextView {
    private Paint paint;
    public CanvasView(Context context) {
        this(context, null);
    }

    public CanvasView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CanvasView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        paint = new Paint();
        paint.setStrokeWidth(20);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setAntiAlias(true);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        canvas.drawCircle(300, 300, 200, paint);

        paint.setColor(Color.RED);
        canvas.drawCircle(300, 300, 150, paint);

        paint.setColor(Color.WHITE);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(2);
        canvas.drawCircle(300, 300, 100, paint);

        paint.setStyle(Paint.Style.FILL);
        paint.setStrokeWidth(20);
        paint.setColor(Color.BLACK);
        float[] points = {50, 50, 50, 100, 100, 50, 100, 100, 150, 50, 150, 100};
        canvas.drawPoints(points, paint);


    }
}

File: app/src/main/java/com/example/harry/customandroid/tabs/develop/customView/CanvasView.kt
package com.example.harry.customandroid.tabs.develop.customView

import android.content.Context
import android.graphics.*
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View
import com.example.harry.customandroid.utils.dp2px

class CanvasView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : View(context, attrs, defStyleAttr) {

    private val paint = Paint(ANTI_ALIAS_FLAG)
    private val textPaint = TextPaint(ANTI_ALIAS_FLAG)
    private val path = Path()

    private val text1 = "Lorem Ipsum is simply dummy text of the printing and typesetting industry."
    private lateinit var staticLayout: StaticLayout

    init {
        paint.textSize = dp2px(20f)
        paint.style = Paint.Style.STROKE

        textPaint.textSize = dp2px(30f)
    }

    companion object {
        const val text = "Hello PQ"
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        path.moveTo(0f, 0f)
        path.lineTo(w / 2f, h.toFloat())
        path.lineTo(w.toFloat(), 0f)

        staticLayout = StaticLayout(text1, textPaint, w, Layout.Alignment.ALIGN_NORMAL,
                1f, 0f, true)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        canvas?.drawPath(path, paint)

        canvas?.drawTextOnPath(text, path, 100f, 0f, textPaint)

        canvas?.drawText(text, 200f, 200f, paint)

        staticLayout.draw(canvas)

    }


}
